Code review n.2 for Turcato and Scialpi (by Enrico Cestaro)

EnricoCestaro opened the pull request:
Hi, I'm asking for three files to be reviewed, one was entirely written by me, one was partially written by me, one was edited and generalized for multiple tasks (by me)
You can ignore, inside the idFinderHandler class (considering the 'green text') the first 4 cases, under the comment /**Joining a network */.

Files to review:
. idFinderHandler
. ResourceExchangeHandler
. SMSKademliaListener
(They contain a description of their functionalities and at the bottom of the ResourceExchangeHandler class I added a formal description about how a resource is added to the network dictionary)

At a first approach, it might seem like a lot of code but, in reality, a lot of it is repeated with a few differences, as for example the methods createAddRequest, createGetRequest, createDeleteRequest. There are a lot of comments and specifications, and it doesn't accomplish any complicated task

Nothing else to say, I hope it will not be problematic, and thank you.

EnricoCestaro committed:
re-adding files for the review

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+         * @throws IllegalArgumentException If the the key or the resource are null or invalid
+         */
+        public Request(String key, String resource) {
+            if (key == null) throw new IllegalArgumentException();
```
Maybe it would be better
```
if (key == null || key.length() == 0)
throw new IllegalArgumentException(MESSAGE);

```

EnricoCestaro replied:
Oh. I thought I already corrected it. My bad

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+     *                 network to allow the final receiver to send the answer right to the initial node
+     * @throws IllegalArgumentException If the idToFind or the searcher are null
+     */
+    public void processRequest(KademliaId idToFind, SMSPeer searcher, ResearchMode researchMode) {
```
This method does `if (searcher == null || idToFind == null || researchMode == null)` which is redundant since this method only calls IdFinderHandler.searchId() which does the same stuff. The whole method is redundant in fact. Maybe you added it because it's easier to understand `processRequest` than `searchId`. So maybe `searchId `should be called in another way.
Or maybe `processRequest ` should handle the job of the various createFooRequest of this class, since all of them have this code inside:
```
        Request currentRequest = new Request(key, resource);
        KademliaId idToFind = currentRequest.getKeyId();
        pendingFooRequests.put(idToFind, currentRequest);
        //Start to search for the closest ID

```
You would avoid some repetition.

The result would be something like 
```
processRequest(Request currentR, ResearchMode mode, SMSPeer searcher){
     KademliaId idToFind = currentR.getKeyId();
     switch(mode){
           //Here you add on the right Map of pending requests
     }
     IdFinderHandler.searchId(idToFind, searcher, mode);
}
```

It should be achievable.
Otherwise this method is completely redundant since `searchId `is public

turcato-niccolo replied:
Anyway take this with a grain of salt, since I didn't invent this logic, but you did, and you would know better than me if this is acceptable

EnricoCestaro replied:
Ok, now I see what you meant; it's a really good idea, Requests are made to accept null resources, so it should be more than enough to specify the ResearchMode to know what kind of request it's being processed.
To be honest, I did the same thing in searchId, and I wasn't sure about the choice of the switch-case for the ResearchMode, but now that I see it I can say that it looks really good. Thank you

turcato-niccolo replied:
Happy to help, sometimes I find myself that to see if a solution is practical I have to write some kind of sketch

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+    public void completeAddRequest(KademliaId idToFind, SMSPeer targetPeer) {
+        if (idToFind == null || targetPeer == null) throw new IllegalArgumentException();
+        //1. Find in the pendingAddRequests the Request to complete, remove it from the list
+        IRequest aboutToClose = pendingAddRequests.get(idToFind);
```
What if `idToFind `is not found in the map? It's not written in the specifications that `idToFind `must be a previously added `KademliaId` for an `AddRequest`, so you would have to handle it in some way.

turcato-niccolo replied:
The same can be said for the other methods `completeFooRequest`, and some of the work that these methods perform could be centralized like I suggested for `processRequest`

EnricoCestaro replied:
Well, the methods are only called by the SMSKademliaListener, so you have to receive a request for a specific resource to find, this was a simplified version of the code which had to handle only basic situations, so I dind't worry about it, but I should have added it in the specification as you say, beacuse yes, if I somehow receive the order to close a request that I never created the program doesn't know what to do

EnricoCestaro replied:
For now, I will simply add a description in the specifications; as you will read in MY code review, you will notice that we have two different styles: you always handle exceptions, I just warn the user about them (when I remember to do it :)  )

Probably your choice is better, we'll see

turcato-niccolo replied:
Nice

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+    public void createAddRequest(String key, String resource) {
+        if(key == null || resource == null) throw new IllegalArgumentException();
+        //Create the Request object, insert it inside the pendingAddRequests list
+        Request currentRequest = new Request(key, resource);
```
These lines are repeated in each `createFooRequest` method, see the comment about `processRequest`

EnricoCestaro replied:
For what concerns the processRequest method I have to say I initially thought it could be a good idea to create a wrapper method inside the ResourceExchangeClass, that in fact simply calls the searchId method, but that also helped me to visualize the different steps of the mechanism to process different requests. It wasn't necessary, and I understand it.
About the repetition of lines, at a first look they can seem completely identical, but you'll notice that:
1. Add requests are initialized with Request(key, resource), while Delete and Get requests with Request(key, null)
2. the list inside which the requests are stored changes upon the type of request (so that it becomes pendingAddRequest.put(...), pendingGetRequest.put(...), pendingDeleteRequest.put(...) )
3. The processRequest method gets a different ResearchMode value based upon the type of Request.

So, as you can see, there are a few differences that held me from creating a single method for all the Requests

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+
+        @Override
+        public String toString() {
+            return "Request_ID: " + this.getKeyId().toString();
```
I don't see how this is relevant as `toString()` you should add a description

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/IdFinderHandler.java:
```diff
@@ -1,2 +1,153 @@
+     */
+    public static void searchId(KademliaId idToFind, SMSPeer searcher, ResearchMode researchMode) {
+        if (searcher == null || idToFind == null || researchMode == null)
+            throw new IllegalArgumentException();
```
You should specify a message, I get that it's launched only if a certain condition occurs and it's well-explained int he specs, but I don't think this is good practice. The same is done in **all the code**. 
For example when debugging it's handy to see an explanatory message with the exception launched so that you don't have to go read the full documentation. And you know just by that what caused the failure.

EnricoCestaro replied:
I think you are totally right, this could be a precious addition to my coding skills, especially for debugging; and yes, it's a problem in all my code, from now it will be a good habitude of mine doing it

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/SMSKademliaListener.java:
```diff
@@ -108,7 +108,138 @@ public void onMessageReceived(SMSMessage message) {
+                sendAcknowledge(peer);
+                //2. Processes the information brought by the message received
+                Log.i("CONN_LOG", "IdFound: " + splitted[1]);
+                idToFind = new KademliaId(splitted[1]);
```
`splitted[1]` -> very much hard-wired, you should at least define somewhere a constant for the index of each parameter to have
`splitted[ID_INDEX]` for example
Better again it would be having a parser for the messages so the logic of the syntax is defined elsewhere and CAN be tested. Doing like this, it's a hell to do testing, and I think it's fair to say that this is NOT a good practice.

EnricoCestaro replied:
Mh, mooving it elsewhere could be a good idea, it will be part of our restoration process (which we already started for all our code) because we noticed (and someone else made us notice) how we stacked too many functionalities in the same spots. About the hardwiring, the position is not always related to the same type of data, for example: sometimes the second value is the searcher peer, sometimes is the resource; but yes, it could be avoided/solved mooving the process elsewhere, even if it would require a different handling mechanism based upon the type of task to accomplish. Because, as I said, the structure of the incoming message changes upon the RequestType; it was an easier solution handling in loco the situation, maybe not the most modular. It will require some time, but it could be a good way to improve the ability of the system to process more complex informations

turcato-niccolo replied:
> About the hardwiring, the position is not always related to the same type of data, for example: sometimes the second value is the searcher peer, sometimes is the resource; but yes, it could be avoided/solved mooving the process elsewhere, even if it would require a different handling mechanism based upon the type of task to accomplish. Because, as I said, the structure of the incoming message changes upon the RequestType; it was an easier solution handling in loco the situation, maybe not the most modular.

`sometimes the second value is the searcher peer, sometimes is the resource` Ok I didn't see that, I must say it's even worse. I understand the practicality, but if in the future you will need to change this code, you'll find it very difficult to understand. Too many cases.
I suggest that you and your team decide a fixed syntax, and add it in the java docs. I think you shouldn't let pass too much time, as you'll find it more difficult to fix it.

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/SMSKademliaListener.java:
```diff
@@ -108,7 +108,138 @@ public void onMessageReceived(SMSMessage message) {
+                //status, that is that I'm alive and still active inside the network
+                sendAcknowledge(peer);
+                //2. Processes the information brought by the message received
+                Log.i("CONN_LOG", "IdFound: " + splitted[1]);
```
I think here you forgot `LOG_TAG`

EnricoCestaro replied:
YEP

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/SMSKademliaListener.java:
```diff
@@ -108,7 +108,138 @@ public void onMessageReceived(SMSMessage message) {
+                ConnectionHandler.acceptRequest(peer);
+                break;
+            case FindId:
+                //1. Creates the response addressed to the sender, to inform him of my activity
```
This comment is repeated each time `sendAcknowledge(peer)` appears, maybe you should move on the top of the switch. Or maybe leave it full length here and in the following appearances write a more brief comment like:
`inform the sender of my alive status`. Or you could even not write it for the others, `Acknowledge(peer)` is quite self-explanatory.

EnricoCestaro replied:
I tried to remove it from each case, unfortunately, as you can see JoinPermission and AcceptJoin do not require the acknowledge message, and I still have to find a better solution than copying it inside almost every case. About the comments, yeah, the name is pretty explicative, I should remember that who is reading already knows I'm writing code for Kademlia

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+     * distributed dictionary, plus the {@link KademliaId} of the resource key, used to distinguish
+     * each Request from the other
+     */
+    public class Request implements IRequest {
```
I don't quite like the name the class `Request`, not very coherent, maybe consider changing it? I hope it doesn't sound silly but maybe `ResourceRequest `would sound better? Since the identifier for the request is a resource key, it's a specific type of request.

EnricoCestaro replied:
Well, it could, but I wanted to avoid names like pendingAddResourceRequests, it was a bit long. So, my reasoning was silly too. I'll think about it

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/ResourceExchangeHandler.java:
```diff
@@ -1,2 +1,310 @@
+     *
+     * @param key The String value of the key of the resource to get from the Dictionary
+     */
+    public void createGetRequest(String key) {
```
Missing `@throws IllegalArgumentException` specification, the same is repeated in other methods.

EnricoCestaro replied:
I will fix it 👍

turcato-niccolo reviewed (commented):
Ok, in summary, the code is nice and modular, it's not quite simple though, for me at least, I haven't done much with pending requests, so I felt a little drunk after reading this. But that's ok.
Anyway, I see that there is some nice and hard thinking behind this code.
I think it's little too much for a single review though, but I'd like to hear an opinion from another reviewer
I think that the code has no formal problem, specifications are up to standards, there are some issues though. And one BIG problem in my opinion:
The message syntax you're using is very much hard-wired: partially because the separator used is a " " (blank space), you should use a constant defined somewhere, but mainly because the syntax is based on a concatenation of strings. So it's not easy to maintain and pretty much impossible to test separately. I would suggest using a parser of some type or defining a builder.

`IdFinderHandler`:
I understood the logic, but with not easily. I think this class is a little weighed down by comments. I mean, I think I could understand the logic without them since all names are relevant and self-explanatory. The same can be said for the rest of the code.

`ResourceExchangeHandler`:
Some repetition of code, that makes the code not so easy to read, without that it would be much clearer.
Some methods are missing specifications.

`SMSKademliaListener`:
There is the result of the messages' syntax hard-wiring, but the logic used in the handlers makes that the managing of requests and results are based on a few lines and it's good.

EnricoCestaro committed:
Custom messages handler created

Upon the reviewer advice, I added the KademliaMessage class, which allows creating custom messages with specific structure and fields, it's been implemented inside the SMSKademliaListener to easily create and read messages from and to the network; all the building process has been moved inside the new class. Also, I modified the ResourceExchangeHandler class, now the methods to create and close Requests are generalized. In the next commit I'll add specifications and descriptions for the methods

EnricoCestaro committed:
Specifications added

EnricoCestaro committed:
Added specifications and custom EXCEPTIONS

EnricoCestaro committed:
I noticed a couple of mistakes

I forgot to convert part of the code to the new structure, now it should be correct

EnricoCestaro commented on webdictionary/src/main/java/com/eis0/kademlianetwork/IdFinderHandler.java:
```diff
@@ -1,2 +1,158 @@
+     * @param targetPeer  The Peer representing the target that will receive the result
+     */
+    private static void sendResult(RequestTypes requestType, KademliaId idToFind, SMSPeer targetPeer) {
+        KademliaMessage kadMessage = new KademliaMessage(requestType, idToFind, null, null, null);
+        SMSMessage searchResult = new SMSMessage(targetPeer, kadMessage.toString());
+        SMSManager.getInstance().sendMessage(searchResult);
```
@turcato-niccolo This is an example of how is created and sent a message with the new structure.
The major add I did was the class KademliaMessage

turcato-niccolo replied:
Nice, much better👍

EnricoCestaro commented on webdictionary/src/main/java/com/eis0/kademlianetwork/SMSKademliaListener.java:
```diff
@@ -63,26 +63,21 @@ public void sendPong(SMSPeer peer){
+        String resource = kadMessage.resource;
+        //The two values, idToFind and idFound, share the same field, depending upon the type of request
+        //the SMSKademliaListener knows which one of the two is occupying it, and process it consequently
+        KademliaId idFound = idToFind;
```
@turcato-niccolo Now the SMSMessage received is converted into a KademliaMessage, where the fields are pre-defined. I made the choice to not make private all the variables inside the KademliaMessage, all the data are brought via SMS and are contained inside the message, which is simply converted into a more readable format; I didn't think it could be necessary to create get<Value> methods for each field, but I wait to know your opinion.

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/KademliaMessage.java:
```diff
@@ -0,0 +1,105 @@
+        text = message.getData();
+        peer = message.getPeer();
+        //Array of strings containing the message fields
+        String[] splitted = text.split(" ");
```
Watch out, this is hard-wiring, keep a constant `separator` instead

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/KademliaMessage.java:
```diff
@@ -0,0 +1,105 @@
+        if (requestType == null) throw new IllegalArgumentException(REQUEST_TYPE_NULL);
+        //The RequestType is the only field which needs to be not Null
+        String stringIdToFind;
+        if (idToFind == null) stringIdToFind = "/";
```
Define a` final String default = "/"` instead

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/KademliaMessage.java:
```diff
@@ -0,0 +1,105 @@
+ * This class is used to create custom messages to send or receive inside the Kademlia Network, the
+ * fields are in common for all the type of requests
+ */
+public class KademliaMessage {
```
A much better solution, but it has the same problems of hardwiring as before. Anyway, you can easily solve them just defining some constants. I think having public fields is fine as long as this class remains private inside your package

EnricoCestaro replied:
Yeah, the positions of the fields are still hardwired, I could change them with some constants, but I'm still not sure if it's worth it, considering that that's the only place where they are used, the method itself is quite small and self-explanatory, and inside the method description there is a list of the position of the parameters inside the message; it takes nothing to change them, but with this new structure I honestly don't feel the need to define those constants anymore.

turcato-niccolo replied:
Count that like this it's not much expandable, it's your choice if it's worth it. The important is that the logic of syntax is well contained in just one point of access.
I would suggest adding a description of the syntax.

EnricoCestaro committed:
Avoiding hardwiring

Added the new final values BLANK = " "; DEFAULT = "/"

turcato-niccolo commented on webdictionary/src/main/java/com/eis0/kademlianetwork/KademliaMessage.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws com.eis.smslibrary.exceptions.InvalidTelephoneNumberException If If telephoneNumber
+     *                                                                       check is not {@link SMSPeer.TelephoneNumberState#TELEPHONE_NUMBER_VALID}.
+     */
+    public KademliaMessage(SMSMessage message) {
```
Since the "output" of this class is a String, maybe consider doing the same with the "input".
Just to be coherent, I would have a constructor `KademliaMessage(String message, Peer sender)`. Or a method `getMessage()` that returns a `SMSMessage ` built from the `toString()`.

EnricoCestaro replied:
I think it would be even better to move the peer handling process outside the KademliaMessage class, because as a value it's easily obtainable; this way, I can let the class process only the text message, as you said, inserted as a String value

turcato-niccolo commented:
Nice, now you'll see that testing this is much easier.

EnricoCestaro committed:
Peer handling process moved

Now the KademliaMessage class receives and returns only the String value of the message itself, the Peer which sent the message is obtained and processed elsewhere

EnricoCestaro committed:
Correction completed

Added the @author, corrected a little mistake in the toString() method
